stack/new_stack.py

```diff
@@ -0,0 +1,176 @@
+            'datasets:customer_id': 'C010ys3gt',
+            'datasets:enable_release': release,
+            'gcp:billing_project': _gcp_project,
+            'gcp:project': dataset,
+            'gcp:user_project_override': 'true',
+            **formed_hail_config,
+        },
+    }
+
+    with open(pulumi_config_fn, 'w+', encoding='utf-8') as fp:
+        print(f'Writing to {pulumi_config_fn}')
+        yaml.dump(pulumi_stack, fp, default_flow_style=False)
+
+    command = f"""\
+    export PULUMI_CONFIG_PASSPHRASE=
+    pulumi stack init {dataset}
+    pulumi stack select {dataset}
+    pulumi up
+    """
+    print(f'Created stack for {dataset}, can deploy with:\n{command}')
+
+
+def create_project(project_id, organisation_id=ORGANIZATION_ID):
+    """Call subprocess.check_output to create project under an organisation"""
+    command = [
+        'gcloud',
+        'projects',
+        'create',
+        project_id,
+        '--organization',
+        organisation_id,
+    ]
+    check_output(command)
+
+
+def assign_billing_account(project_id, billing_account_id=BILLING_ACCOUNT_ID):
+    """
+    Assign a billing account to a GCP project
+    """
+    command = [
+        'gcloud',
+        'beta',
+        'billing',
+        'projects',
+        'link',
+        project_id,
+        '--billing-account',
+        billing_account_id,
+    ]
+    check_output(command)
+
+
+def create_budget(project_id: str, amount=100):
+    """
+    Create a monthly budget for the project_id
+    """
+    command = [
+        'gcloud',
+        *('--project', BILLING_PROJECT_ID),
+        *('billing', 'budgets', 'create'),
+        *('--display-name', project_id),
+        *('--billing-account', BILLING_ACCOUNT_ID),
+        *('--filter-projects', f'projects/{project_id}'),
+        *('--budget-amount', f'{amount:.2f}AUD'),
+        *('--calendar-period', 'month'),
+        *(
+            '--threshold-rule=percent=0.5',
+            '--threshold-rule=percent=0.9',
+            '--threshold-rule=percent=1.0',
+        ),
+        *(
+            '--notifications-rule-pubsub-topic',
+            f'projects/{BILLING_PROJECT_ID}/topics/budget-notifications',
+        ),
+    ]
+
+    return check_output(command)
+
+
+def get_hail_service_accounts(dataset: str):
+    """Get hail service accounts from kubectl"""
+    check_output(
+        [
+            'gcloud',
+            f'--project={HAIL_PROJECT}',
+            'container',
+            'clusters',
+            'get-credentials',
+            'vdc',
+        ]
+    )
+    hail_tokens = check_output(GET_HAIL_TOKENS(dataset), shell=True).decode()
+    hail_client_emails_by_level = get_client_emails_from_kubectl_output(hail_tokens)
+
+    return hail_client_emails_by_level
+
+
+def get_client_emails_from_kubectl_output(output):
+    """
+    Kubectl gives us 3 json blobs, so use a regex to find the hail emails
+    """
+    responses = RE_CLIENT_EMAIL_MATCHER.findall(output)
```

Comment:
That's a bit hacky, but parsing the json is kind of painful, I know...

Reply:
The problem is it returns 3 json blobs, which I couldn't think of a trivial way to work out how to pick up each blob, without seeking and counting braces.

Reply:
But that's just because `GET_HAIL_TOKENS` runs a for-loop? Couldn't you instead pull out the loop, so you can parse this more easily?

Reply:
Yeah that's a much better idea, I've changed it to do this instead